src/modules/samples/cpp/src/so/SampleModule.cpp

```diff
@@ -0,0 +1,270 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <cstdio>
+#include <cstring>
+
+#include <Sample.h>
+#include <ScopeGuard.h>
+#include <Mmi.h>
+
+void __attribute__((constructor)) InitModule()
+{
+    SampleLog::OpenLog();
+    OsConfigLogInfo(SampleLog::Get(), "C++ Sample module loaded");
+}
+
+void __attribute__((destructor)) DestroyModule()
+{
+    OsConfigLogInfo(SampleLog::Get(), "C++ Sample module unloaded");
+    SampleLog::CloseLog();
+}
+
+int MmiGetInfo(
+    const char* clientName,
+    MMI_JSON_STRING* payload,
+    int* payloadSizeBytes)
+{
+    int status = MMI_OK;
+
+    constexpr const char ret[] = R""""({
+        "Name": "C++ Sample",
+        "Description": "A sample module written in C++",
+        "Manufacturer": "Microsoft",
+        "VersionMajor": 1,
+        "VersionMinor": 0,
+        "VersionInfo": "",
+        "Components": ["SampleComponent"],
+        "Lifetime": 1,
+        "UserAccount": 0})"""";
+
+    ScopeGuard sg{[&]()
+    {
+        if (MMI_OK == status)
+        {
+            if (IsFullLoggingEnabled())
+            {
+                OsConfigLogInfo(SampleLog::Get(), "MmiGetInfo(%s, %.*s, %d) returned %d", clientName, *payloadSizeBytes, *payload, *payloadSizeBytes, status);
+            }
+            else
+            {
+                OsConfigLogInfo(SampleLog::Get(), "MmiGetInfo(%s, -, %d) returned %d", clientName, *payloadSizeBytes, status);
+            }
+        }
+        else
+        {
+            if (IsFullLoggingEnabled())
+            {
+                OsConfigLogError(SampleLog::Get(), "MmiGetInfo(%s, %.*s, %d) returned %d", clientName, *payloadSizeBytes, *payload, *payloadSizeBytes, status);
+            }
+            else
+            {
+                OsConfigLogError(SampleLog::Get(), "MmiGetInfo(%s, -, %d) returned %d", clientName, *payloadSizeBytes, status);
+            }
+        }
+    }};
+
+    if (nullptr == clientName)
+    {
+        OsConfigLogError(SampleLog::Get(), "MmiGetInfo called with null clientName");
+        status = EINVAL;
+    }
+    else if (nullptr == payload)
+    {
+        OsConfigLogError(SampleLog::Get(), "MmiGetInfo called with null payload");
+        status = EINVAL;
+    }
+    else if (nullptr == payloadSizeBytes)
+    {
+        OsConfigLogError(SampleLog::Get(), "MmiGetInfo called with null payloadSizeBytes");
+        status = EINVAL;
+    }
+    else
```

Comment:
As commented on the Sample.h, this could be a static method in the base class (GetInfo) that does this there in the static lib also to be exposed to unittesting.